### ISSUE TITLE
pbsim2 simulation of HiFi reads annotated by truth positions using vg inject

### ISSUE THREAD

**Issue by StephenHwang** [vgteam/vg#3666]
**1. What were you trying to do?**
Convert pbsim2 simulated HiFi reads to a gam format annotated with truth read position.

**2. What did you want to happen?**
Get a gam of simulated reads annotated by the truth position from the graph.

**3. What actually happened?**
An error that the alignment references path are not present in the graph.

**4. If you got a line like `Stack trace path: /somewhere/on/your/computer/stacktrace.txt`, please copy-paste the contents of that file here:**

Did not receive a stacktrace; received an error
```
Error: [vg::alignment.cpp] error: alignment references path not present in graph: _thread_NA19239_1_0_0
```

**5. What data and command can the vg dev team use to make the problem happen?**

Run vg paths
```
vg paths \
    --xg /nanopore/cgl/data/giraffe/mapping/graphs/for-NA19239/1000gplons/hs38d1/1000GPlons_hs38d1.xg \
    --gbwt /nanopore/cgl/data/giraffe/mapping/graphs/for-NA19239/1000gplons/hs38d1/1000GPlons_hs38d1.gbwt \
    --sample NA19239 \
    --extract-fasta \
  > /public/groups/vg/sjhwang/vg_scripts/bin/reads/sim_HiFi_other_tools/sim_pbsim2/NA19239_paths.fa
```

Create a GBZ
```
/public/groups/vg/sjhwang/vg_full/bin/vg_retagging gbwt \
  -g graph.gbz --gbz-format -x /nanopore/cgl/data/giraffe/mapping/graphs/for-NA19239/1000gplons/hs38d1/1000GPlons_hs38d1.xg \
  /nanopore/cgl/data/giraffe/mapping/graphs/for-NA19239/1000gplons/hs38d1/1000GPlons_hs38d1.gbwt
```

Retag GBZ ( see: https://github.com/vgteam/vg/pull/3658 )
```
/public/groups/vg/sjhwang/vg_full/bin/vg_retagging gbwt -Z \
  --set-tag "reference_samples=NA19239" \
  --gbz-format -g /public/groups/vg/sjhwang/data/rf/NA19239/graphs/retagged.gbz \
  /public/groups/vg/sjhwang/data/rf/NA19239/graphs/graph.gbz
```

Run pbsim2
```
pbsim --depth 20 \
       --accuracy-min 0.00 \
       --length-min 10000 \
       --difference-ratio 6:50:54 \
       --sample-fastq /public/groups/vg/sjhwang/data/reads/real_HiFi/tmp/HiFi_reads_100k_real.fq \
       --prefix sim_NA19239/chr \
       /public/groups/vg/sjhwang/vg_scripts/bin/reads/sim_HiFi_other_tools/sim_pbsim2/NA19239_paths.fa \
       &> tmp.log
```
Output: For each read in fastq output from vg paths (.vg) we get an output: maf, fastq, and ref file
Located: /public/groups/vg/sjhwang/vg_scripts/bin/reads/sim_HiFi_other_tools/sim_pbsim2/sim_NA19239/pbsim_sim_output/chr1_head

Convert maf to sam
```
bioconvert maf2sam \
      /public/groups/vg/sjhwang/vg_scripts/bin/reads/sim_HiFi_other_tools/sim_pbsim2/sim_NA19239/maf_to_sam_files/example/example.maf \
      /public/groups/vg/sjhwang/vg_scripts/bin/reads/sim_HiFi_other_tools/sim_pbsim2/sim_NA19239/maf_to_sam_files/example/example.sam            # maf to sam
```
Note: then in the sam file, rename the reference name from "`ref`" to the corresponding reference read name; ie. "`_thread_NA19239_1_0_0`".
Example files located: /public/groups/vg/sjhwang/vg_scripts/bin/reads/sim_HiFi_other_tools/sim_pbsim2/sim_NA19239/maf_to_sam_files/example

vg inject using retagged gbz (previously before retagging `--xg-name /nanopore/cgl/data/giraffe/mapping/graphs/for-NA19239/1000gplons/hs38d1/1000GPlons_hs38d1.xg`)
```
/public/groups/vg/sjhwang/vg_full/bin/vg_retagging inject \
  --xg-name /public/groups/vg/sjhwang/data/rf/NA19239/graphs/retagged.gbz \
  /public/groups/vg/sjhwang/vg_scripts/bin/reads/sim_HiFi_other_tools/sim_pbsim2/sim_NA19239/maf_to_sam_files/example/example.sam > \
  /public/groups/vg/sjhwang/vg_scripts/bin/reads/sim_HiFi_other_tools/sim_pbsim2/sim_NA19239/maf_to_sam_files/example/example.gam
```

**6. What does running `vg version` say?**

```
vg version v1.40.0-80-g0b6ac5914 "Suardi"
Compiled with g++ (GCC) 11.2.0 on Linux
Linked against libstd++ 20210728
Built by sjhwang@courtyard
```

**Comment by adamnovak** [comment]
Your problem here is that your `vg` and your `vg_retagging` are separated by when I merged #3646, which means they use different string names to talk about the paths stored in the GBWT and the GBZ.

What used to be called `_thread_NA19239_1_0_0` (sample NA19239, contig "1", haplotype 0, fragment 0) is now called `NA19239#0#1#0` if it's a haplotype (sample NA19239, haplotype 0, contig "1", fragment 0; note the changed order), and `NA19239#0#1` if you change it from a haplotype to a reference (because we don't have multiple fragments for reference paths so we can use proper [PanSN names](https://github.com/pangenome/PanSN-spec)).

So you should be able to rename your SAM file to use  `NA19239#0#1` instead of `_thread_NA19239_1_0_0` as the contig name and then inject will work.

```
(venv) [anovak@courtyard toil]$ cat  /public/groups/vg/sjhwang/vg_scripts/bin/reads/sim_HiFi_other_tools/sim_pbsim2/sim_NA19239/maf_to_sam_files/example/example.sam | head -n4 | sed 's/_thread_NA19239_1_0_0/NA19239#0#1/g' >part.sam
(venv) [anovak@courtyard toil]$ vg inject --xg-name /public/groups/vg/sjhwang/data/rf/NA19239/graphs/retagged.gbz part.sam >part.gam

(venv) [anovak@courtyard toil]$ echo $?
0
(venv) [anovak@courtyard toil]$ vg stats -a part.gam
Total alignments: 1
Total primary: 1
Total secondary: 0
Total aligned: 0
Total perfect: 0
Total gapless (softclips allowed): 0
Total paired: 0
Total properly paired: 0
Insertions: 5 bp in 5 read events
Deletions: 3 bp in 1 read events
Substitutions: 0 bp in 0 read events
Softclips: 0 bp in 0 read events
```

Except you would have to call `vg_retagging` if you aren't merging everything into one `vg` build.

So, in conclusion **just run your SAM through `sed 's/_thread_NA19239_1_0_0/NA19239#0#1/g'` and it will magically start working**.

**Comment by StephenHwang** [comment]
That solved it. Thank you!
